bftengine/src/preprocessor/PreProcessor.cpp

```diff
@@ -287,35 +321,64 @@ void PreProcessor::onMessage<ClientPreProcessRequestMsg>(ClientPreProcessRequest
       return;
     }
-    seqNumberOfLastReply = myReplica_.seqNumberOfLastReplyToClient(clientId);
-    if (seqNumberOfLastReply == reqSeqNum) {
+    const bool replySentToClient = myReplica_.isReplyAlreadySentToClient(clientId, reqSeqNum);
+    if (replySentToClient) {
       LOG_INFO(logger(),
                "Request has already been executed - let replica to decide how to proceed further"
                    << KVLOG(reqSeqNum, clientId));
       preProcessorMetrics_.preProcReqSentForFurtherProcessing.Get().Inc();
-      return incomingMsgsStorage_->pushExternalMsg(clientPreProcessReqMsg->convertToClientRequestMsg(false));
+      incomingMsgsStorage_->pushExternalMsg(move(clientMsg));
+      return;
     }
-    if (seqNumberOfLastReply < reqSeqNum)
-      registerSucceeded = registerReplicaDependentRequest(
-          move(clientPreProcessReqMsg), preProcessRequestMsg, ++(clientEntry->reqRetryId));
+    if (myReplica_.isCurrentPrimary())
+      registerSucceeded = registerRequestOnPrimaryReplica(
+          move(clientMsg), preProcessRequestMsg, msgOffsetInBatch, (reqEntry->reqRetryId)++);
+    else
+      registerAndHandleClientPreProcessReqOnNonPrimary(move(clientMsg), arrivedInBatch, msgOffsetInBatch);
   }
-
-  if (myReplica_.isCurrentPrimary() && registerSucceeded) {
-    preProcessorMetrics_.preProcInFlyRequestsNum.Get().Inc();  // Increase this metric on the primary replica
+  if (myReplica_.isCurrentPrimary() && registerSucceeded)
     return handleClientPreProcessRequestByPrimary(preProcessRequestMsg);
-  }
 
   LOG_DEBUG(logger(),
             "ClientPreProcessRequestMsg" << KVLOG(reqSeqNum, clientId, senderId)
                                          << " is ignored because request is old/duplicated");
   preProcessorMetrics_.preProcReqIgnored.Get().Inc();
-}  // namespace preprocessor
+}
+
+template <>
+void PreProcessor::onMessage<ClientBatchRequestMsg>(ClientBatchRequestMsg *msg) {
+  ClientBatchRequestMsgUniquePtr clientBatch(msg);
+  LOG_DEBUG(logger(),
+            "Received ClientBatchPreProcessRequestMsg"
+                << KVLOG(clientBatch->clientId(), clientBatch->getCid(), clientBatch->numOfMessagesInBatch()));
+  if (!checkClientBatchMsgCorrectness(clientBatch)) {
+    preProcessorMetrics_.preProcReqIgnored.Get().Inc();
+    return;
+  }
+  ClientMsgsList &clientMsgs = clientBatch->getClientPreProcessRequestMsgs();
+  uint16_t offset = 0;
+  for (auto &clientMsg : clientMsgs) {
+    LOG_DEBUG(logger(),
+              "Start handling single message from the batch:" << KVLOG(clientMsg->requestSeqNum(),
+                                                                       clientMsg->clientProxyId(),
+                                                                       clientMsg->senderId(),
+                                                                       clientMsg->requestTimeoutMilli()));
+    handleSingleClientRequestMessage(move(clientMsg), true, offset++);
+  }
+  if (!myReplica_.isCurrentPrimary()) {
```

Comment:
Shouldn’t this check happen in the beginning of the method?

Reply:
This message could be received by either primary replica or not, both cases are valid.